### ISSUE TITLE
Local Outlier Factor for Anomaly Detection

### ISSUE THREAD

**Issue by iefymenkoanna** [ZDDduesseldorf/audioexplorer-backend#41]
## The first test version 

[audio-anomaly-detection.zip](https://github.com/user-attachments/files/28341512/audio-anomaly-detection.zip)

The actual version is on GitHub.

## Summary

Implement Local Outlier Factor (LOF) to calculate a local anomaly score for each audio embedding. The algorithm should be applied after the embedding vectors have been generated.

This issue documents the implementation of anomaly detection using Local Outlier Factor on audio embeddings.

The implementation operates directly on an embedding dictionary and does not require API integration.

The anomaly detection is executed through the `calculate_anomalies()` function and returns anomaly scores together with human-readable labels.

---

## Goal

Detect locally unusual audio samples based on their embeddings.

Calculate anomaly scores for each audio sample by comparing them to their nearest neighbors.

The implementation should:

* receive embeddings as input
* execute Local Outlier Factor on all embeddings
* calculate anomaly scores
* convert scores into anomaly labels
* return structured results for each audio sample

---

## Description

Each audio sample is represented as a numerical embedding vector.

Local Outlier Factor (LOF) analyzes the local density around each embedding and compares it to the density of neighboring embeddings.

Samples that have significantly lower density than their surrounding neighbors receive higher anomaly scores.

The resulting score should be normalized and displayed in the Web UI as a percentage.

---

## Input Structure

The function receives a dictionary containing embeddings.

Example:

```python
{
    "uuid1": [...],
    "uuid2": [...],
    "uuid3": [...]
}
```

Where:

* key = audio UUID
* value = embedding vector

---

## Implementation

Main entry point:

```python
calculate_anomalies()
```

Workflow:

1. Receive embedding dictionary
2. Extract embedding vectors
3. Execute Local Outlier Factor
4. Calculate anomaly scores
5. Normalize scores
6. Convert scores into labels
7. Return structured result dictionary

---

## Output Structure

The anomaly detection service returns a dictionary containing scores and labels for each audio sample.

Example:

```python
{
    "uuid1": {
        "scores": {
            "isolation_forest": 34.87,
            "lof": 78.77
        },
        "labels": {
            "isolation_forest": "Slightly Anomalous",
            "lof": "Anomalous"
        }
    }
}
```

---

## sklearn

We use a density-based anomaly detection algorithm.

The implementation uses:

```python
from sklearn.neighbors import LocalOutlierFactor
```

Local Outlier Factor compares the local density of a sample with the densities of its nearest neighbors.

Samples located in significantly less dense regions are considered anomalous.

LOF is especially useful for detecting local anomalies that may not be detected by global methods such as Isolation Forest.

---

## Configuration

We need important configurations for our algorithm.

### "n_neighbors"

* Number of neighboring samples used for density comparison.
* Controls how local the anomaly detection should be.
* Smaller values focus on local structures.
* Larger values consider a broader neighborhood.

### "contamination"

* Controls how difficult it is for a data point to be classified as an anomaly.
* Represents the expected proportion of anomalies in the dataset.
* We could assume that approximately 3% of samples are anomalous.

### "metric"

* Distance metric used to compare embeddings.
* Common choice: Euclidean distance.
* Determines how similarity between embeddings is calculated.

---

## Label Assignment

Raw scores are converted into human-readable labels
(based on their normalized anomaly score).

Labels: 

+ Not Anomalous (0-30%)
+ Slightly Anomalous (30-50%)
+ Anomalous (50-80%)
+ Highly Anomalous (80-100%)

Responsibilities:

* interpret anomaly scores
* assign anomaly categories
* provide consistent labels

---

## Unit Tests

Tests verify:

* processing of embedding dictionaries
* LOF execution
* anomaly score calculation
* score normalization
* label assignment
* output structure validation
* edge cases and boundary values

---

## Pipeline Position

Audio
→ Feature Extraction
→ CNN / Audio Encoder
→ Embeddings (numeric vectors)
→ Isolation Forest
→ LOF
→ Anomaly Score (%), Anomaly Label
→ UMAP
→ Web UI

---

## UI Requirements

Display the score in the Web UI using the algorithm name:

LOF: 78.77% (example)

Label: Anomalous (example)

Optional:

* Tooltip: What is a Local Outlier Factor algorithm?
* Tooltip text should explain that this is an unsupervised algorithm.
* Diagrams (Bar chart).

---

## Notes

* Works directly on embedding vectors.
* Does not require labels for the input.
* Detects local anomalies.
* Unsupervised learning approach.
* Complements Isolation Forest by focusing on neighborhood density.

---

## Acceptance Criteria

* [x] Local Outlier Factor integrated
* [x] Embedding dictionary accepted as input
* [x] Scores calculated successfully
* [x] Labels assigned correctly
* [x] Output structure matches specification
* [x] Unit tests passing
* [x] Ruff checks passing
* [x] MyPy checks passing
* [x] CI pipeline passing
